File: csharp/Patterns/Mvp.cs
using System;

namespace DesignPatterns.Patterns;

public static class Mvp
{
    private class CounterModel
    {
        public int Value { get; private set; }

        public void Increment() => Value++;
    }

    private interface ICounterView
    {
        void SetPresenter(CounterPresenter presenter);
        void Render(int value);
    }

    private class CounterView : ICounterView
    {
        private CounterPresenter? _presenter;

        public void SetPresenter(CounterPresenter presenter)
        {
            _presenter = presenter;
        }

        public void ClickIncrement()
        {
            _presenter?.IncrementClicked();
        }

        public void Render(int value)
        {
            Console.WriteLine($"View shows: {value}");
        }
    }

    private class CounterPresenter
    {
        private readonly CounterModel _model;
        private readonly ICounterView _view;

        public CounterPresenter(CounterModel model, ICounterView view)
        {
            _model = model;
            _view = view;
            _view.SetPresenter(this);
        }

        public void IncrementClicked()
        {
            _model.Increment();
            _view.Render(_model.Value);
        }
    }

    public static void Run()
    {
        Console.WriteLine("\nModel View Presenter");
        var model = new CounterModel();
        var view = new CounterView();
        _ = new CounterPresenter(model, view);
        view.ClickIncrement();
        view.ClickIncrement();
    }
}

File: python/patterns/mvp.py
class CounterModel:
    def __init__(self) -> None:
        self.value = 0

    def increment(self) -> None:
        self.value += 1


class CounterView:
    def __init__(self) -> None:
        self.presenter: "CounterPresenter" | None = None

    def set_presenter(self, presenter: "CounterPresenter") -> None:
        self.presenter = presenter

    def click_increment(self) -> None:
        if self.presenter:
            self.presenter.increment_clicked()

    def render(self, value: int) -> None:
        print(f"View shows: {value}")


class CounterPresenter:
    def __init__(self, model: CounterModel, view: CounterView) -> None:
        self.model = model
        self.view = view
        self.view.set_presenter(self)

    def increment_clicked(self) -> None:
        self.model.increment()
        self.view.render(self.model.value)


def run() -> None:
    print("\nModel View Presenter")
    model = CounterModel()
    view = CounterView()
    CounterPresenter(model, view)
    view.click_increment()
    view.click_increment()
